File: internal/engine/sstable_builder.go
package engine

import (
	"bytes"
	"encoding/binary"
	"os"
)

const blockSize = 4096

type IndexEntry struct {
	Key    []byte
	Offset uint32
}

type SSTableBuilder struct {
	file              *os.File
	offset            uint32
	blockBytesWritten uint32
	index             []IndexEntry

	minKey []byte
	maxKey []byte
	bloom  *BloomFilter

	// FIX: Reusable scratch buffer to prevent GC pressure during high-velocity writes
	scratchBuf []byte
}

func NewSSTableBuilder(path string) (*SSTableBuilder, error) {
	f, err := os.OpenFile(path, os.O_CREATE|os.O_WRONLY|os.O_TRUNC, 0644)
	if err != nil {
		return nil, err
	}

	return &SSTableBuilder{
		file:       f,
		index:      make([]IndexEntry, 0, 1000),
		bloom:      NewBloomFilter(4096, 3),
		scratchBuf: make([]byte, 1024), // Start with a 1KB buffer
	}, nil
}

func (b *SSTableBuilder) Add(internalKey []byte, value []byte) error {
	userKey, _, _ := ParseInternalKey(internalKey)

	// Update Min/Max using UserKey only
	if b.minKey == nil || bytes.Compare(userKey, b.minKey) < 0 {
		b.minKey = append([]byte(nil), userKey...)
	}
	if b.maxKey == nil || bytes.Compare(userKey, b.maxKey) > 0 {
		b.maxKey = append([]byte(nil), userKey...)
	}

	b.bloom.Add(string(userKey))

	if b.blockBytesWritten == 0 {
		b.index = append(b.index, IndexEntry{
			Key:    append([]byte(nil), userKey...),
			Offset: b.offset,
		})
	}

	// 1. Calculate record size
	recordSize := 4 + len(internalKey) + 4 + len(value)

	// 2. REUSE OR GROW the scratch buffer instead of allocating new memory
	if cap(b.scratchBuf) < recordSize {
		// If we hit a massive value, grow the buffer.
		// Go's slice growth will ensure this happens infrequently.
		b.scratchBuf = make([]byte, recordSize)
	}
	// Slice the buffer to the exact size needed
	buf := b.scratchBuf[:recordSize]

	// 3. Serialize directly into the scratch memory
	binary.LittleEndian.PutUint32(buf[0:4], uint32(len(internalKey)))
	offset := 4

	copy(buf[offset:], internalKey)
	offset += len(internalKey)

	binary.LittleEndian.PutUint32(buf[offset:offset+4], uint32(len(value)))
	offset += 4

	copy(buf[offset:], value)

	// 4. Single system call write
	_, err := b.file.Write(buf)
	if err != nil {
		return err
	}

	b.offset += uint32(recordSize)
	b.blockBytesWritten += uint32(recordSize)

	if b.blockBytesWritten >= blockSize {
		b.blockBytesWritten = 0
	}

	return nil
}

func (b *SSTableBuilder) Finish() error {
	// ... (Rest of the logic for Bloom, Index, and Footer remains as previously established)
	// These are called only once per SSTable, so the single-time allocations here are acceptable.

	bloomStartOffset := b.offset
	if _, err := b.file.Write(b.bloom.Bytes()); err != nil {
		return err
	}
	b.offset += uint32(len(b.bloom.Bytes()))

	indexStartOffset := b.offset

	countBuf := make([]byte, 4)
	binary.LittleEndian.PutUint32(countBuf, uint32(len(b.index)))
	if _, err := b.file.Write(countBuf); err != nil {
		return err
	}
	b.offset += 4

	for _, entry := range b.index {
		entrySize := 4 + len(entry.Key) + 4
		entryBuf := b.scratchBuf[:entrySize] // Reuse scratchBuf here too!

		binary.LittleEndian.PutUint32(entryBuf[0:4], uint32(len(entry.Key)))
		copy(entryBuf[4:], entry.Key)
		binary.LittleEndian.PutUint32(entryBuf[4+len(entry.Key):], entry.Offset)

		if _, err := b.file.Write(entryBuf); err != nil {
			return err
		}
		b.offset += uint32(entrySize)
	}

	metaStartOffset := b.offset
	metaSize := 4 + len(b.minKey) + 4 + len(b.maxKey)
	metaBuf := b.scratchBuf[:metaSize] // Reuse scratchBuf here too!

	binary.LittleEndian.PutUint32(metaBuf[0:4], uint32(len(b.minKey)))
	metaOffset := 4
	copy(metaBuf[metaOffset:], b.minKey)
	metaOffset += len(b.minKey)

	binary.LittleEndian.PutUint32(metaBuf[metaOffset:metaOffset+4], uint32(len(b.maxKey)))
	metaOffset += 4
	copy(metaBuf[metaOffset:], b.maxKey)

	if _, err := b.file.Write(metaBuf); err != nil {
		return err
	}
	b.offset += uint32(metaSize)

	footerBuf := make([]byte, 16)
	binary.LittleEndian.PutUint32(footerBuf[0:4], bloomStartOffset)
	binary.LittleEndian.PutUint32(footerBuf[4:8], indexStartOffset)
	binary.LittleEndian.PutUint32(footerBuf[8:12], metaStartOffset)
	binary.LittleEndian.PutUint32(footerBuf[12:16], 0xABCD1234)

	if _, err := b.file.Write(footerBuf); err != nil {
		return err
	}

	b.file.Sync()
	return b.file.Close()
}

File: internal/engine2_0/sstable_builder.cpp
#include "sstable_builder.hpp"
#include "internal_key.hpp"
#include <algorithm>
#include <stdexcept>

static void PutUint32LE(uint8_t* buf, uint32_t val) {
    buf[0] = static_cast<uint8_t>(val);
    buf[1] = static_cast<uint8_t>(val >> 8);
    buf[2] = static_cast<uint8_t>(val >> 16);
    buf[3] = static_cast<uint8_t>(val >> 24);
}

SSTableBuilder::SSTableBuilder(const std::string& path)
    : offset_(0),
      blockBytesWritten_(0),
      bloom_(std::make_unique<BloomFilter>(4096, 3)),
      scratchBuf_(1024) {
    file_.open(path, std::ios::out | std::ios::binary | std::ios::trunc);
    if (!file_.is_open()) {
        throw std::runtime_error("failed to open sstable file for writing: " + path);
    }
}

SSTableBuilder::~SSTableBuilder() {
    if (file_.is_open()) {
        file_.close();
    }
}

bool SSTableBuilder::Add(const std::vector<uint8_t>& internalKey, const std::vector<uint8_t>& value) {
    std::vector<uint8_t> userKey;
    uint64_t seqNum = 0;
    uint8_t keyType = 0;
    if (!ParseInternalKey(internalKey, userKey, seqNum, keyType)) {
        return false;
    }

    if (minKey_.empty() || userKey < minKey_) {
        minKey_ = userKey;
    }
    if (maxKey_.empty() || userKey > maxKey_) {
        maxKey_ = userKey;
    }

    bloom_->Add(std::string(userKey.begin(), userKey.end()));

    if (blockBytesWritten_ == 0) {
        index_.push_back({userKey, offset_});
    }

    size_t recordSize = 4 + internalKey.size() + 4 + value.size();
    if (scratchBuf_.size() < recordSize) {
        scratchBuf_.resize(recordSize);
    }

    PutUint32LE(scratchBuf_.data(), static_cast<uint32_t>(internalKey.size()));
    size_t offset = 4;

    std::copy(internalKey.begin(), internalKey.end(), scratchBuf_.data() + offset);
    offset += internalKey.size();

    PutUint32LE(scratchBuf_.data() + offset, static_cast<uint32_t>(value.size()));
    offset += 4;

    if (!value.empty()) {
        std::copy(value.begin(), value.end(), scratchBuf_.data() + offset);
    }

    if (!file_.write(reinterpret_cast<const char*>(scratchBuf_.data()), recordSize)) {
        return false;
    }

    offset_ += static_cast<uint32_t>(recordSize);
    blockBytesWritten_ += static_cast<uint32_t>(recordSize);

    if (blockBytesWritten_ >= blockSize) {
        blockBytesWritten_ = 0;
    }

    return true;
}

bool SSTableBuilder::Finish() {
    uint32_t bloomStartOffset = offset_;
    const auto& bloomBytes = bloom_->Bytes();
    if (!file_.write(reinterpret_cast<const char*>(bloomBytes.data()), bloomBytes.size())) {
        return false;
    }
    offset_ += static_cast<uint32_t>(bloomBytes.size());

    uint32_t indexStartOffset = offset_;
    uint8_t countBuf[4];
    PutUint32LE(countBuf, static_cast<uint32_t>(index_.size()));
    if (!file_.write(reinterpret_cast<const char*>(countBuf), 4)) {
        return false;
    }
    offset_ += 4;

    for (const auto& entry : index_) {
        size_t entrySize = 4 + entry.Key.size() + 4;
        if (scratchBuf_.size() < entrySize) {
            scratchBuf_.resize(entrySize);
        }

        PutUint32LE(scratchBuf_.data(), static_cast<uint32_t>(entry.Key.size()));
        std::copy(entry.Key.begin(), entry.Key.end(), scratchBuf_.data() + 4);
        PutUint32LE(scratchBuf_.data() + 4 + entry.Key.size(), entry.Offset);

        if (!file_.write(reinterpret_cast<const char*>(scratchBuf_.data()), entrySize)) {
            return false;
        }
        offset_ += static_cast<uint32_t>(entrySize);
    }

    uint32_t metaStartOffset = offset_;
    size_t metaSize = 4 + minKey_.size() + 4 + maxKey_.size();
    if (scratchBuf_.size() < metaSize) {
        scratchBuf_.resize(metaSize);
    }

    PutUint32LE(scratchBuf_.data(), static_cast<uint32_t>(minKey_.size()));
    size_t metaOffset = 4;
    std::copy(minKey_.begin(), minKey_.end(), scratchBuf_.data() + metaOffset);
    metaOffset += minKey_.size();

    PutUint32LE(scratchBuf_.data() + metaOffset, static_cast<uint32_t>(maxKey_.size()));
    metaOffset += 4;
    std::copy(maxKey_.begin(), maxKey_.end(), scratchBuf_.data() + metaOffset);

    if (!file_.write(reinterpret_cast<const char*>(scratchBuf_.data()), metaSize)) {
        return false;
    }
    offset_ += static_cast<uint32_t>(metaSize);

    uint8_t footerBuf[16];
    PutUint32LE(footerBuf, bloomStartOffset);
    PutUint32LE(footerBuf + 4, indexStartOffset);
    PutUint32LE(footerBuf + 8, metaStartOffset);
    PutUint32LE(footerBuf + 12, 0xABCD1234U);

    if (!file_.write(reinterpret_cast<const char*>(footerBuf), 16)) {
        return false;
    }

    file_.flush();
    file_.close();
    return true;
}
